File: frontend/src/lib/schemas/cluster.ts
import { z } from 'zod';

import { tickerResponseSchema } from '@/lib/schemas/ticker';

export const clusterCreateSchema = z.object({
  name: z.string().min(1).max(100),
  /** Hex colour, exactly 7 chars, e.g. '#4a90d9'. */
  color: z
    .string()
    .length(7)
    .regex(/^#[0-9A-Fa-f]{6}$/, 'Must be a valid hex colour (#RRGGBB)'),
});

export const clusterUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  color: z
    .string()
    .length(7)
    .regex(/^#[0-9A-Fa-f]{6}$/)
    .optional(),
});

export const clusterResponseSchema = z.object({
  id: z.number(),
  name: z.string(),
  color: z.string(),
  tickers: z.array(tickerResponseSchema).default([]),
  created_at: z.string(),
  updated_at: z.string(),
});

export type ClusterCreate = z.infer<typeof clusterCreateSchema>;
export type ClusterUpdate = z.infer<typeof clusterUpdateSchema>;
export type ClusterResponse = z.infer<typeof clusterResponseSchema>;

File: backend/src/atlas/schemas/cluster.py
"""Pydantic schemas for position clusters."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from atlas.schemas.ticker import TickerResponse


class ClusterCreate(BaseModel):
    """Payload to create a new cluster."""

    name: str = Field(min_length=1, max_length=100)
    # Hex colour string — exactly 7 chars, e.g. "#4a90d9".
    color: str = Field(
        min_length=7,
        max_length=7,
        pattern=r"^#[0-9A-Fa-f]{6}$",
        description="Hex colour, e.g. '#4a90d9'.",
    )


class ClusterUpdate(BaseModel):
    """Payload to update a cluster (all fields optional)."""

    name: str | None = Field(None, min_length=1, max_length=100)
    color: str | None = Field(
        None,
        min_length=7,
        max_length=7,
        pattern=r"^#[0-9A-Fa-f]{6}$",
    )


class ClusterResponse(BaseModel):
    """API response for a cluster, including its assigned tickers."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    color: str
    tickers: list[TickerResponse] = []
    created_at: datetime
    updated_at: datetime
